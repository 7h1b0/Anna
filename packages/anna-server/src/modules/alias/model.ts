import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import aliasSchema from './schema';
import { omit } from '../../utils/utils';

export const TABLE = 'alias';
export const COLUMNS = [
  'aliasId',
  'sceneId',
  'name',
  'description',
  'enabled',
  'startTime',
  'endTime',
  'createdAt',
  'updatedAt',
  'createdBy',
];

type Alias = {
  aliasId: string;
  sceneId: string;
  name: string;
  description?: string;
  startTime?: number;
  endTime?: number;
  enabled: boolean;
  createdAt: number;
  updatedAt: number;
  createdBy: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(aliasSchema, data);
}

export async function findAll(): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(aliasId: string): Promise<Alias> {
  return knex(TABLE).first(COLUMNS).where('aliasId', aliasId);
}

export async function findByName(name: string): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS).where('name', name);
}

export async function save({
  sceneId,
  name,
  description,
  createdBy,
  enabled = true,
  startTime,
  endTime,
}: {
  sceneId: string;
  name: string;
  description: string;
  createdBy: string;
  enabled?: boolean;
  startTime?: number;
  endTime?: number;
}): Promise<string> {
  const aliasId = uuidv4();
  await knex(TABLE).insert({
    aliasId,
    sceneId,
    description,
    name,
    enabled,
    createdBy,
    startTime,
    endTime,
  });
  return aliasId;
}

export function remove(aliasId: string) {
  return knex(TABLE).where('aliasId', aliasId).del();
}

export function findByIdAndUpdate(aliasId: string, payload: Partial<Alias>) {
  const safePayload = omit(
    payload,
    'aliasId',
    'createdAt',
    'updatedAt',
    'createdBy',
  );
  return knex(TABLE).update(safePayload).where('aliasId', aliasId);
}

export function isInteger(value: unknown | number): value is number {
  return Number.isInteger(value);
}

/**
 * Given aliases, returns one alias respecting the constraints
 */
export function resolveActiveAlias(aliases: Alias[]): Alias | undefined {
  const time = new Date().getHours();
  const activeAlias = aliases
    .filter((alias) => alias.enabled)
    .filter((alias) => {
      if (isInteger(alias.startTime) && isInteger(alias.endTime)) {
        if (alias.startTime < alias.endTime) {
          return time >= alias.startTime && time < alias.endTime;
        }
        return time >= alias.startTime || time < alias.endTime;
      }
      return true;
    });

  return activeAlias.pop();
}
