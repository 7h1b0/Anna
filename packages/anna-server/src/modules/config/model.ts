import knex from '../../knexClient';

export const TABLE = 'config';

export type Config = {
  isAway: boolean;
};

export async function fetchIsAway(): Promise<{ isAway: true }> {
  return knex(TABLE).first('isAway');
}

export function updateIsAway(isAway: boolean) {
  return knex(TABLE).update({ isAway });
}
