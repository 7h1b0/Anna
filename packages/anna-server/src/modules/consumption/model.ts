import knex from '../../knexClient';

export const TABLE = 'consumption';
export const COLUMNS = ['date', 'value'];

export type Consumption = {
  date: number;
  value: number;
};

export async function findLastEntries(limit: number): Promise<Consumption[]> {
  return knex(TABLE)
    .select(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(limit);
}

export async function findLastEntry(): Promise<Consumption> {
  return knex(TABLE)
    .first(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(1);
}

export async function save(consumptions: { date: Date; value: number }[]) {
  return knex.transaction((trx) => {
    return Promise.all(
      consumptions.map((consumption) => trx.insert(consumption).into(TABLE)),
    );
  });
}
