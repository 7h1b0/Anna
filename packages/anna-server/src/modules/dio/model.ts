import Ajv from 'ajv';
import knex from '../../knexClient';
import dioSchema from './schema';

export const TABLE = 'dios';

export type Dio = {
  dioId: number;
  roomId: string;
  name: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE).first('dioId', 'roomId', 'name').where('dioId', dioId);
}

export async function findByRoomId(roomId: string): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name').where('roomId', roomId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE).where('dioId', dioId).del();
}

export function findByIdAndUpdate(
  dioId: number,
  payload: Partial<Omit<Dio, 'dioId'>>,
) {
  return knex(TABLE).update(payload).where('dioId', dioId);
}
