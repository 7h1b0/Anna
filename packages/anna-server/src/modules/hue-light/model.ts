import knex from '../../knexClient';
export const TABLE = 'lights';
export const COLUMNS = ['roomId', 'lightId'];

export function findRoomId(lightId: number): Promise<string[]> {
  return knex(TABLE).first('roomId').where('lightId', lightId);
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function findByRoomId(roomId: string): Promise<number[]> {
  return knex(TABLE)
    .select('lightId')
    .where('roomId', roomId)
    .then((res) => res.map(({ lightId }) => lightId));
}

export function remove(lightId: number): Promise<number> {
  return knex(TABLE).where('lightId', lightId).del();
}

export async function insertOrUpdate(lightId: number, roomId: string) {
  await knex.raw(
    'INSERT INTO lights (lightId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [lightId, roomId],
  );
}
