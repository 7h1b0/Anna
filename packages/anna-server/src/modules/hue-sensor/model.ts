import Ajv from 'ajv';
import knex from '../../knexClient';
import sensorSchema from './schema';

export const TABLE = 'sensors';
export const COLUMNS = ['roomId', 'sensorId'];

export function findByRoomId(roomId: string): Promise<string[]> {
  return knex(TABLE)
    .select('sensorId')
    .where('roomId', roomId)
    .then((res) => res.map(({ sensorId }) => sensorId));
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function remove(sensorId: string): Promise<number> {
  return knex(TABLE).where('sensorId', sensorId).del();
}

export async function insertOrUpdate(sensorId: string, roomId: string) {
  await knex.raw(
    'INSERT INTO sensors (sensorId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [sensorId, roomId],
  );
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(sensorSchema, data);
}
