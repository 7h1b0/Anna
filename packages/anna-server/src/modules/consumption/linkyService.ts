import linky from '@bokub/linky';
import isBefore from 'date-fns/isBefore';
import addDays from 'date-fns/addDays';
import format from 'date-fns/format';

// import { schedule } from 'services/scheduleService';
import { findLastEntry, save } from 'modules/consumption/model';
import * as logger from 'utils/logger';

const LINKY_USER = process.env.LINKY_USER;
const LINKY_PASSWORD = process.env.LINKY_PASSWORD;

function formatDate(date: Date): string {
  return format(date, 'dd/MM/yyyy');
}

export async function getRange() {
  const yesterday = addDays(new Date(), -1);

  const lastEntry = await findLastEntry();
  if (!lastEntry) {
    const lastWeek = addDays(yesterday, -7);

    return {
      start: formatDate(lastWeek),
      end: formatDate(yesterday),
    };
  }

  const dayAfterLastEntry = addDays(new Date(lastEntry.date), 1);

  if (isBefore(dayAfterLastEntry, yesterday)) {
    return {
      start: formatDate(dayAfterLastEntry),
      end: formatDate(yesterday),
    };
  }
  return null;
}

export async function fetchLinkyData() {
  try {
    logger.info('Retrieving Linky value');
    const range = await getRange();

    if (range) {
      const session = await linky.login(LINKY_USER, LINKY_PASSWORD);
      logger.info(JSON.stringify(range));
      const data: {
        date: string;
        value: number;
      }[] = await session.getDailyData(range);

      await save(
        data.map((consumption) => ({
          date: new Date(consumption.date),
          value: consumption.value,
        })),
      );
    } else {
      logger.info(`Last entries is yesterday`);
    }
  } catch (err) {
    logger.error(`Error while retrieving Linky value: ${err.message}`);
  }
}

export function run() {
  logger.info('Linky service disabled');
  // schedule('linky', '0 0 10 * * *', fetchLinkyData);
}
