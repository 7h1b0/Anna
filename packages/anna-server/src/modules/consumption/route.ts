import { Router } from 'express';
import * as Consumption from './model';

const routes = Router();
routes.route('/api/consumption').get((req, res) => {
  Consumption.findLastWeek()
    .then((consumptions) => res.json(consumptions))
    .catch((err) => res.status(500).send({ err }));
});

export default routes;
