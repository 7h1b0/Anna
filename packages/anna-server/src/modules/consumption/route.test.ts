import request from 'supertest';
import * as lolex from '@sinonjs/fake-timers';
import { createUser } from 'factories';
import knex from 'knexClient';
import * as Consumption from 'modules/consumption/model';
import * as User from 'modules/user/model';
import app from '../../index';

const user = createUser();
const seventeenth = {
  date: new Date(Date.UTC(2020, 7, 17)),
  value: 7.0,
};
const twentieth = {
  date: new Date(Date.UTC(2020, 7, 20)),
  value: 5.01,
};
const twentyFirst = {
  date: new Date(Date.UTC(2020, 7, 21)),
  value: 5.01,
};
const initConsumptions = [seventeenth, twentieth, twentyFirst];

describe('Consumption API', () => {
  beforeAll(async () => {
    await knex(User.TABLE).truncate();
    await knex(User.TABLE).insert(user);
  });

  beforeEach(async () => {
    await knex(Consumption.TABLE).truncate();
    await knex(Consumption.TABLE).insert(initConsumptions);
  });

  describe('/api/consumption', () => {
    describe('GET', () => {
      it('should return all consumptions for the last 7 days', async () => {
        const clock = lolex.install({ now: new Date('2020-08-25T16:00') });
        const response = await request(app)
          .get('/api/consumption')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);
        clock.uninstall();

        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/consumption')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });
    });
  });
});
