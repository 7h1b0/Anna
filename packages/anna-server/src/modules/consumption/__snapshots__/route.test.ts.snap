// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consumption API /api/consumption GET should return all consumptions for the last 7 days 1`] = `
Array [
  Object {
    "date": 1597968000000,
    "value": 5.01,
  },
  Object {
    "date": 1597881600000,
    "value": 5.01,
  },
  Object {
    "date": 1597622400000,
    "value": 7,
  },
]
`;
